### PR TITLE
feat: Keycloak service as wrapper

### PR DESCRIPTION
**Pull request by wtrocki**
## Idea

It should be able to compose services:
We have now 4 different services:

- AMQ Service (for proper topic and mapping)
- Keylcoak Service
- DataSync Service
- CRUDService (SubscriptionService)

DataSync-starter needs to use them all 🙂
there is no ability to compose services (as it is also important for services to be executed in the single order.  Only one instance of the service can exist.

Keycloak service and AMQ services should be comparable - means that they can accept both:

- DataSync Service
- CRUDService (SubscriptionService)

This capability do not require any API changes and can chain our creator pattern (which is simple enough for people to get understanding)

## Progress

- [x] Keycloak Wrapper service
- [ ] Tests
- [ ] Export Keycloak configuration to the documentation (as auth)

### CODE REVIEW

**Comment by wtrocki** [comment]
Keycloak lib is using outdated API. Going to change it, but looking for early feedback.

File: packages/graphback-keycloak-authz/src/KeycloakCrudService.ts [hunk 1]
```diff
@@ -29,13 +29,14 @@ export type KeycloakCrudServiceOptions = {
  * This is actually quite nice and clean but it does not allow for field level auth.
  * It's still a possibility that we could go with though!
  */
-export class KeycloakCrudService<T = any> extends CRUDService<T> {
+export class KeycloakCrudService<T = any> implements GraphbackCRUDService<T> {
 
   private authConfig: CrudServiceAuthConfig;
+  private wrappingService: GraphbackCRUDService;
```
Comment (craicoverflow) [inline packages/graphback-keycloak-authz/src/KeycloakCrudService.ts:35]: Isn't this (KeycloakCrudService) technically the wrapping service?

File: docs/authentication/keycloak.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+id: keycloak
+title: Out of the box Keycloak based authentication
+sidebar_label: Keycloak Auth
+---
+
+## Graphback Keycloak Authz
+
+Graphback Keycloak Authz enables [Keycloak](https://www.keycloak.org/) integration in [Graphback](https://graphback.dev) based applications. This enables you to declaratively add authorization capabilities like role bassed access on top of the CRUD model that is used within Graphback.
+
+The `KeycloakCrudService` extends Graphback's default `CrudService` class which provides the authorization capabilities. See [the docs](https://graphback.dev/docs/crud/crudruntime) for more information about the Crud model in Graphback.
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply docs/authentication/keycloak.md:11]: Ignore entire keycloak file. Making code changes to get that sorted.

**Comment by wtrocki** [comment]
Ready for review

File: docs/authentication/introduction.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+---
+
+Graphback abstracts from the authentication mechanism.
+By default Graphback templates come without authentication, but it is easy to retrofit any existing authentication mechanism 
+on top of Graphback.
+
+With the `GraphbackCRUDService` abstraction developers can add any rules and implement various different authorization policies.
+Graphback developers recomend following libraries:
+
+- Passport.js if you looking for in application authorization and authentication mechanisms.
```
Comment (machi1990) [inline docs/authentication/introduction.md:14]:
```suggestion
- [Passport.js](http://www.passportjs.org/) if you looking for in application authorization and authentication mechanisms.
```

File: docs/authentication/introduction.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+- Keycloak SSO - Standalone SSO server offering integration with various social OAuth logins and federating users.
```
Comment (machi1990) [inline docs/authentication/introduction.md:15]:
```suggestion
- [Keycloak](https://www.keycloak.org/) SSO - Standalone SSO server offering integration with various social OAuth logins and federating users.
```

File: packages/graphback-core/src/index.ts [hunk 5]
```diff
@@ -8,12 +8,15 @@ export * from './plugin/GraphbackGlobalConfig'
...
 export * from './relationships/relationshipHelpers';
-export * from './runtime'
-export * from './db';
 export * from './annotations/DefaultValueAnnotation';
 
+export * from './utils/printSchemaWithDirectives';
+export * from './utils/metadataAnnotations';
+export * from './utils/fieldTransformHelpers';
+
+export * from './runtime'
```
Comment (machi1990) [inline packages/graphback-core/src/index.ts:21]:
```suggestion
export * from './runtime';
```

File: packages/graphback-keycloak-authz/README.md [hunk 6]
```diff
@@ -2,287 +2,9 @@
...
   <img width="400" src="https://raw.githubusercontent.com/aerogear/graphback/master/website/static/img/logo.png">
   <br/>
-  Auto generate database structure, <br/>
-  GraphQL Resolvers and Queries from GraphQL types 🚀
+  RBAC Authorization on top of the Graphback Service layer<br/>
+  Declarative code and schema driven RBAC for Graphback 🚀
 </p>
 
-**Documentation**: https://graphback.dev
+**Documentation**: http://localhost:3000/docs/next/authentication/keycloak
```
Comment (machi1990) [inline packages/graphback-keycloak-authz/README.md:9]:
```suggestion
**Documentation**: https://graphback.dev/docs/next/authentication/keycloak
```

File: docs/authentication/keycloak.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+const server = new ApolloServer({
+  typeDefs: printSchema(schema),
+  resolvers,
+  context: (req) => {
+    return {
+      kauth: new KeycloakContext({ req })
+    }
+  }
+})
+```
```
[the remark this answers is not in the document]
Comment (machi1990) [reply docs/authentication/keycloak.md:121]: Yeap, it should use the new runtime api

File: packages/graphback-core/src/index.ts (hunk 5 above)
Comment (machi1990) [reply packages/graphback-core/src/index.ts:21]: Noticed that sometime we add a semicolon, sometimes we don't. This an be applied or ignored as you see fit but I think we should be consistent.

File: docs/authentication/keycloak.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (wtrocki) [reply docs/authentication/keycloak.md:121]: Outdated comment 😆

[the remark this answers is not in the document]
Comment (wtrocki) [reply docs/authentication/keycloak.md:121]: Actually I lost this change :(

File: packages/graphback-keycloak-authz/README.md (hunk 6 above)
Comment (wtrocki) [reply packages/graphback-keycloak-authz/README.md:9]: Yes.. Sorry - I have lost one commit during rebase :(

File: packages/graphback-core/src/index.ts (hunk 5 above)
Comment (wtrocki) [reply packages/graphback-core/src/index.ts:21]: when is vscode it adds it but devs do not add it :)
I really hate eslint package. It is so convoluted and it doesn't format code right. 
Using prettier will be much better but using both at the same time gets tricky as styles need to align. I will create issue to see if we can reformat all code using prettier and then make it pass using lint

Comment (machi1990) [reply packages/graphback-core/src/index.ts]: Cool, thanks.

File: docs/authentication/keycloak.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+With this configuration the following rules are in place.
+
+- All users can create and read `Task` types but only admins can update and delete them.
+- Admin users can create, update and delete `Report` types, and all users can read them.
+
+## Example with Graphback Runtime
+
+During server initialization, use the `createKeycloakRuntimeContext` function to initialize the KeycloakCrudService instances for each model.
```
Comment (machi1990) [inline docs/authentication/keycloak.md:55]:
```suggestion
During server initialization, use the `createKeycloakCRUDService` function to initialize the KeycloakCrudService instances for each model.
```

File: packages/graphback-keycloak-authz/src/createKeycloakCRUDService.ts [hunk 9]
```diff
@@ -1,36 +1,19 @@
...
-  /**
-   * PubSub implementation for creating subscriptions
-   */
-  pubSub: PubSubEngine
-  /**
-   * Auth config for entire data model
-   */
-  authConfig: CrudServiceAuthConfig
-}
+import { ModelDefinition, GraphbackDataProvider, GraphbackCRUDService, CreateCRUDServiceOptions, CRUDServiceConfig, CRUDService, ServiceCreator, createCRUDService } from '@graphback/core';
```
Comment (machi1990) [inline packages/graphback-keycloak-authz/src/createKeycloakCRUDService.ts:1]:
```suggestion
import { ModelDefinition, GraphbackDataProvider, GraphbackCRUDService, ServiceCreator } from '@graphback/core';
```

**Comment by wtrocki** [comment]
Amazing review! Thank you so much! Sorry for blocking you with this change. It should be the last big one after we merge this (hopefully)

**Comment by wtrocki** [comment]
@craicoverflow I merged based on @machi1990 review to unblock his work and try to test this changes. Hope you do not mind

**Pull request merged**